sr_speech_control/src/sr_speech_control/speech_control.py

```diff
@@ -15,66 +15,46 @@
 # with this program. If not, see <http://www.gnu.org/licenses/>.
 
 import rospy
-import time
 import speech_recognition as sr
-from difflib import get_close_matches
 from std_msgs.msg import String
 
 
 class SpeechControl(object):
-    def __init__(self, trigger_word, command_words, command_topic='speech_control',
-                 non_speaking_duration=0.1, pause_threshold=0.2):
+    def __init__(self, trigger_word, command_topic='sr_speech_control'):
+        self.microphone = sr.Microphone()
+        for i, microphone_name in enumerate(sr.Microphone.list_microphone_names()):
+            if "pulse" in microphone_name:
+                self.microphone = sr.Microphone(device_index=i)
+                rospy.loginfo("Using microphone: {}".format(microphone_name))
         self.trigger_word = trigger_word
-        self.command_words = command_words
         self.recognizer = sr.Recognizer()
-        self.microphone = sr.Microphone()
         self.command_publisher = rospy.Publisher(command_topic, String, queue_size=1)
-        self.command_to_be_executed = None
-
-        self._init_recognizer(non_speaking_duration, pause_threshold)
         self._stop_listening = self.recognizer.listen_in_background(self.microphone, self._recognizer_callback)
 
-    def _init_recognizer(self, non_speaking_duration, pause_threshold):
-        with self.microphone as source:
-            self.recognizer.adjust_for_ambient_noise(source)
-        self.recognizer.non_speaking_duration = non_speaking_duration
-        self.recognizer.pause_threshold = pause_threshold
-
     def _recognizer_callback(self, recognizer, audio):
         try:
             result = recognizer.recognize_google(audio)
+            rospy.loginfo("Recognition result: {}".format(result))
         except sr.UnknownValueError:
             return
         except sr.RequestError as e:
             rospy.logwarn("Could not request results from Google Speech Recognition service: {}".format(e))
             return
 
-        result = [str(x).lower() for x in result.split(' ')]
-        if self._filter_word(result[0], self.trigger_word) == self.trigger_word:
-            command = self._filter_word(result[1], self.command_words)
-            if command in self.command_words:
-                self.command_to_be_executed = command
-
-    def _filter_word(self, word, dictionary, offset=0.5):
-        result = get_close_matches(word, dictionary, 1, offset)
-        if not result:
-            return word
-        return result[0]
+        result = result.lower()
+        if result.startswith(self.trigger_word) and len(result) > len(self.trigger_word) + 1:
+            self.command_publisher.publish(result[len(self.trigger_word) + 1:])
 
     def run(self):
```

Comment:
If nothing here is really happening, is it possible to just replace this method with `rospy.spin()`?

Reply:
Good idea. Replaced in 0562f5fa3f12391aaefd535d77f96f30363b164b.